feat: Add NATS blobstore provider

ossfellow opened the pull request:
This commit adds a new blobstore provider that uses NATS JetStream as its backend storage. The provider implements the wasi:blobstore/blobstore interface and includes:

- Access to the full range of NATS Object Store Storage features, including persistent and ephemeral storage, and self-expiring objects
- Configurable write timeout to accommodate different computing environments
- Comprehensive integration tests using testcontainers
- Updated http-blobstore example to demonstrate usage of all the blobstore features

The nats-blobstore-provider codebase has been broken down into multiple files for improved maintenance, and to help new contributors understand the structure and the constituent parts of a wasmCloud provider. It also makes it easy to understand which WASI interfaces are implemented by the provider. 

## Feature or Problem
Implementing WASI blobstore proposal by taking advantage of NATS capabilities.

## Related Issues
Addresses #3262 

## Release Information
`next` release

## Consumer Impact
None

## Testing
The provider includes unit tests, for configurations, and a comprehensive integration test suite. However, the integration tests are not running successfully, most likely because of setup issues; all tests have the `ignore` flag set to prevent any adverse effects; so, this should not prevent merging the PR.

To compensate, the `http-blobstore` rust example component has been completely overhauled to test all WASI blobstore functions, which has been used to validate the NATS blobstore provider functionality as well.

It will provide great value to identify the setup issues in the integration test suite, for future use cases, outside the this PR lifecycle.

### Unit Test(s)
All passed

### Acceptance or Integration
Please see Testing.

### Manual Verification
Using the `http-blobstore` example component

brooksmtownsend commented on crates/provider-blobstore-nats/src/blobstore.rs:
```diff
@@ -0,0 +1,645 @@
+
+            let result: Result<(), String> = {
+                // Iterate over the list of objects and delete each object using the delete_object method
+                for object in objects {
+                    self.delete_object(
+                        context.clone(),
+                        bindings::wrpc::blobstore::types::ObjectId {
+                            container: name.clone(),
+                            object: object.clone(),
+                        },
+                    )
+                    .await
+                    .map_err(|err| err.to_string())??; // Handle both levels of Result
+                }
```
This might be a good time to collect these `delete_object` calls into a `Vec<JoinHandle<_>>` and then use https://docs.rs/futures/latest/futures/future/fn.join_all.html to poll all the results concurrently. That might cause problems with NATS object store if it's only expecting one delete at a time, but otherwise I think it'll significantly improve the performance of this function

ossfellow replied:
They say listen to the advice of a seasoned Rustacean 😄 
I took the advice and implemented the deletion in batches of 50; NATS manages concurrent operations gracefully, but it won't hurt to batch the deletions.

brooksmtownsend commented on crates/provider-blobstore-nats/src/config.rs:
```diff
@@ -0,0 +1,443 @@
+    #[serde(default)]
+    pub storage_config: Option<StorageConfig>,
+
+    /// Write operation timeout configuration
+    pub max_write_wait: Option<u64>,
```
Should this be marked as `#[serde(default)]`?

ossfellow replied:
Good observation; thank you. I added it.

brooksmtownsend commented on crates/provider-blobstore-nats/src/config.rs:
```diff
@@ -0,0 +1,443 @@
+/// - `compression` (optional): whether the underlying stream should be compressed; defaults to false
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct StorageConfig {
+    /// Maximum age of any object in the container, expressed in nanoseconds
+    pub max_age: Option<core::time::Duration>,
+    /// Maximum size of the object store container, expressed in bytes
+    pub max_bytes: Option<i64>,
+    /// The type of storage backend, File (default) and Memory
+    pub storage_type: Option<StorageType>,
+    /// How many replicas to keep for each object in a cluster, maximum 5
+    pub num_replicas: Option<usize>,
+    /// Whether the underlying stream should be compressed
+    pub compression: Option<bool>,
```
Same comment re: `#[serde(default)]`

ossfellow replied:
Added to make sure consistency is maintained.

brooksmtownsend commented on crates/provider-blobstore-nats/src/provider.rs:
```diff
@@ -0,0 +1,397 @@
+            opts = add_tls_ca(&ca, opts)?;
+        }
+
+        // Get the cluster_uri
+        let uri = cfg.cluster_uri.unwrap_or_default();
```
By default will this be an empty string or `nats://127.0.0.1:4222`?

ossfellow replied:
Thank you; you did find a bug sir; fixed it by adding the default value from `DEFAULT_NATS_URI`

brooksmtownsend commented on crates/provider-blobstore-nats/src/provider.rs:
```diff
@@ -0,0 +1,397 @@
+            context.as_ref().and_then(|Context { component, .. }| {
+                // Clone component and provide a placeholder default value for link_name if it's None,
+                // which will be replaced by the last link name set by the consumer.
+                let link_name = context
+                    .clone()
+                    .context("missing context")
+                    .unwrap()
+                    .link_name()
+                    .to_string();
```
Are you able to pull the `link_name` from the `Context` object here to avoid the unwrap? I know it's safe because you're already in the code path where `context` is Some, but figured it may help readability.

ossfellow replied:
Agreed; I sometimes forget to take a step back and reevaluate the new improvements and the older code, as a whole; I simplified it.

brooksmtownsend reviewed (commented):
Really fantastic work here @ossfellow ! I still need to review the integration tests file, but your provider and component look great, and you've done a great job with making sure functions are instrumented/traces propagated / all the logic looks really nice. Just a couple comments for the meantime, but I don't see any big blockers here and it looks good to me

ossfellow commented on crates/provider-blobstore-nats/src/blobstore.rs:
```diff
@@ -0,0 +1,645 @@
+        context: Option<Context>,
+        name: String,
+    ) -> anyhow::Result<Result<(), String>> {
+        Ok(async {
```
The pattern was borrowed from the Azure blobstore provider, which seemed to reflect the latest approach in implementing wRPC functions; however, I can think of the following rationale:

- The return types are `anyhow::Result<Result<T, String>>`; the approach helps separate the inner (NATS operations) and outer (provider) level errors, and their handling.
- Scope and context handling.

brooksmtownsend replied:
Yep I'm a fan of this as an easy way to ensure all errors in the function are converted to a String instead of calling `.map_err()` everywhere

ossfellow commented:
@brooksmtownsend, thank you for taking the time to review the PR, and all the great feedback provided. I'd like to mention a few things for your consideration:

1. I used the `Azure blobstore provider` - which seemed to be the most up to date - as the basis of the implementation techniques and approach used; however, the `NATS blobstore provider` has more than a dozen improvements in areas of logic, efficiency and performance over that making it our flagship blobstore provider, and the basis for improving the other 3 blobstore providers.
2. I took your recommendation and implemented concurrent object deletion in batches of `50`; however, I'm not sure if it's a good idea to make it user configurable - similar to the max write buffer for optimized operations in different environments - or it would be an overkill. If you think making it configurable provides a meaningful performance advantage, I'll add that, and I'm open to do it as part of this PR (i.e. tonight), or later, if we're trying to clear the PR backlog before the holidays.
3. The integration test suite is also based on Azure's; but it also has a lot of improvements and documentation; however, as reported before, the wRPC calls fail. If you look at the link setting, you'd see that I set the `wrpc-blobstore/blobstore` interface, but the error messages, which [I shared on our Slack channel](https://wasmcloud.slack.com/archives/C014GEXGUKY/p1733458129403659) allude to `wasi-blobstore/blobstore` being served here; this might point to an issue in our testing framework.

brooksmtownsend commented:
Hey @ossfellow just a quick ask here, would you mind rebasing this PR off of `main` to help get rid of the commits that are from upstream and not a part of this PR? Let me know if this is complicated and if we can help

ossfellow commented:
> Hey @ossfellow just a quick ask here, would you mind rebasing this PR off of `main` to help get rid of the commits that are from upstream and not a part of this PR? Let me know if this is complicated and if we can help

@brooksmtownsend , done; please let me know if anything's amiss

brooksmtownsend commented on crates/provider-blobstore-nats/src/config.rs:
```diff
@@ -0,0 +1,449 @@
+/// - `num_replicas` (optional): how many replicas to keep for each object in a cluster, maximum 5; defaults to 1
+/// - `compression` (optional): whether the underlying stream should be compressed; defaults to false
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct StorageConfig {
+    /// Maximum age of any object in the container, expressed in nanoseconds
+    #[serde(default)]
+    pub max_age: Option<core::time::Duration>,
+    /// Maximum size of the object store container, expressed in bytes
+    #[serde(default)]
+    pub max_bytes: Option<i64>,
+    /// The type of storage backend, File (default) and Memory
+    #[serde(default)]
+    pub storage_type: Option<StorageType>,
+    /// How many replicas to keep for each object in a cluster, maximum 5
+    #[serde(default)]
+    pub num_replicas: Option<usize>,
+    /// Whether the underlying stream should be compressed
+    #[serde(default)]
+    pub compression: Option<bool>,
+}
```
To cut down on `unwrap`s in the codebase, I think it would be better to make all of these fields _required_ but provide the reasonable defaults as you've done below for the `Default` implementation. If you keep the `#[serde(default)]` attribute on each field it should use the default value if it's not provided in configuration

ossfellow replied:
Done; however, I had modelled it after how provider configurations are done.

brooksmtownsend commented on crates/provider-blobstore-nats/src/config.rs:
```diff
@@ -0,0 +1,449 @@
+impl Default for StorageConfig {
+    fn default() -> StorageConfig {
+        StorageConfig {
+            max_age: Some(core::time::Duration::from_secs(315_569_520)), // 10 years
+            max_bytes: Some(1_000_000_000),                              // 1 GB
```
Were these values that you found somewhere, or just reasonable defaults? I could see the argument here to set the max_bytes to `-1` given that's what NATS does (unlimited)

ossfellow replied:
They had remained from the early versions; now they're set to -1

brooksmtownsend commented on crates/provider-blobstore-nats/src/config.rs:
```diff
@@ -0,0 +1,449 @@
+    }
+}
+
+// Performing various provider configuration tests
+#[cfg(test)]
+mod test {
```
Thank you for the unit tests on configuration 🙏🏻

brooksmtownsend commented on crates/provider-blobstore-nats/src/provider.rs:
```diff
@@ -0,0 +1,391 @@
+    /// Build a [`NatsBlobstoreProvider`] from [`HostData`]
+    pub fn from_host_data(host_data: &HostData) -> NatsBlobstoreProvider {
+        let config = NatsConnectionConfig::from_link_config(&host_data.config, &host_data.secrets);
+        if let Ok(config) = config {
+            NatsBlobstoreProvider {
+                default_config: config,
+                ..Default::default()
+            }
```
The tiniest nit, mostly fun syntactic sugar. 
```suggestion
        if let Ok(default_config) = config {
            NatsBlobstoreProvider {
                default_config,
                ..Default::default()
            }
```

brooksmtownsend commented on crates/provider-blobstore-nats/src/provider.rs:
```diff
@@ -0,0 +1,391 @@
+            opts = add_tls_ca(&ca, opts)?;
+        }
+
+        // Get the cluster_uri with proper default
+        let uri = cfg.cluster_uri.unwrap_or(DEFAULT_NATS_URI.to_string());
```
Clippy might yell at you here
```suggestion
        // Get the cluster_uri with proper default
        let uri = cfg.cluster_uri.unwrap_or_else(|| DEFAULT_NATS_URI.to_string());
```

brooksmtownsend commented on crates/provider-blobstore-nats/src/provider.rs:
```diff
@@ -0,0 +1,391 @@
+
+    /// Helper function to lookup and return the NATS JetStream connection handle, and container storage
+    /// configuration, using the client component's context.
+    /// This ensures consistent implementation across all functions that need to get the NATS Blobstore.
+    pub(crate) async fn get_blobstore(
```
```suggestion
    /// This ensures consistent implementation across all functions that need to get the NATS Blobstore.
    #[instrument(level = "debug", skip_all)]
    pub(crate) async fn get_blobstore(
```

Small suggestion to ensure this function is traced, let me know if you left it out purposefully though

brooksmtownsend commented on crates/provider-blobstore-nats/src/provider.rs:
```diff
@@ -0,0 +1,391 @@
+            let components = self.consumer_components.read().await;
+            let nats_stores = components
+                .get(&component_id)
+                .ok_or_else(|| anyhow!("consumer component not linked: {}", component_id))?;
+
+            // Get the NATS Object Store handle and its storage configuration
+            nats_stores
+                .get(&link_name)
+                .cloned()
+                .ok_or_else(|| anyhow!("no NATS Object Store found for link name: {}", &link_name))
```
One more suggestion which is really just syntactic sugar since it aligns with our code style

```suggestion
                .ok_or_else(|| anyhow!("consumer component not linked: {component_id}"))?;

            // Get the NATS Object Store handle and its storage configuration
            nats_stores
                .get(&link_name)
                .cloned()
                .ok_or_else(|| anyhow!("no NATS Object Store found for link name: {link_name}"))
```

brooksmtownsend commented on crates/provider-blobstore-nats/src/provider.rs:
```diff
@@ -0,0 +1,391 @@
+    /// This ensures consistent implementation across all functions that need to list objects.
+    /// Delegates to the core implementation in blobstore.rs.
+    #[instrument(level = "debug", skip_all)]
+    async fn list_container_objects(
```
I think this function and the other functions in this module that call `bindings::exports::wrpc::...` are leftover from a refactor and can be removed, is that right?

brooksmtownsend replied:
That applies for list_objects, delete_object, and delete_objects

brooksmtownsend commented on crates/provider-blobstore-nats/src/provider.rs:
```diff
@@ -0,0 +1,391 @@
+                }
+            }
+        };
+        debug!("NATS Blobstore provider configuration: {:?}", config);
```
sanity check: Does this configuration include any secrets (like the jwt/seed) which we may pull out of secrets and then expose here? If so, we may want to implement the `Display` trait for this type to avoid printing those out

brooksmtownsend commented on crates/provider-blobstore-nats/src/provider.rs:
```diff
@@ -0,0 +1,391 @@
+            Ok(b) => b,
+            Err(e) => {
+                error!("failed to connect to NATS: {:?}", e);
+                bail!(anyhow!(e).context("failed to connect to NATS"))
```
```suggestion
                bail!(e).context("failed to connect to NATS"))
```

This just looked a bit funky, does this construction work?

ossfellow replied:
It doesn't work without the `anyhow!` macro

brooksmtownsend commented on examples/rust/components/http-blobstore/wadm.yaml:
```diff
@@ -79,14 +79,14 @@ spec:
         #
-        # image: file://../../../../crates/provider-blobstore-fs/provider-blobstore-fs.par.gz
-        image: ghcr.io/wasmcloud/blobstore-fs:0.10.2
+        # image: file://../../../../src/bin/blobstore-nats-provider/build/blobstore-nats-provider.par.gz
+        image: ghcr.io/wasmcloud/blobstore-nats:0.1.0
```
note to self: once this merges we should release this to ensure the example keeps working 😄

ossfellow replied:
@brooksmtownsend, reminding you of the reminder 😂

brooksmtownsend commented on examples/rust/components/http-blobstore/wadm.yaml:
```diff
@@ -20,7 +20,12 @@ spec:
-        image: ghcr.io/wasmcloud/components/http-blobstore-rust:0.2.0
+        image: ghcr.io/wasmcloud/http-blobstore-rust:0.3.0
+        id: http-blobstore
+        # config:
+        #   - name: http-blobstore-example
+        #     properties:
+        #       container_names: "ying,yang"
```
nit: if this configuration is unused we can remove it from the YAML

ossfellow replied:
Removed; I had left them to show users how to pass container names.

brooksmtownsend commented on examples/rust/components/http-blobstore/local-wadm.yaml:
```diff
@@ -0,0 +1,93 @@
+        # To use the locally compiled code in this folder, use the line below instead after running `wash build`:
+        image: file://./build/http_blobstore_s.wasm
+        id: http-blobstore
+        # config:
+        #   - name: http-blobstore-example
+        #     properties:
+        #       container_names: "ying,yang"
```
same comment re: unused config

brooksmtownsend commented on crates/provider-blobstore-nats/src/blobstore.rs:
```diff
@@ -0,0 +1,678 @@
+        .map_err(|err| format!("{err:#}")))
+    }
+
+    // Move an object from one to the same (different key name, , or revision number update), or another NATS Blobstore Container
```
```suggestion
    // Move an object from one to the same (different key name, or revision number update), or another NATS Blobstore Container
```

supernit

brooksmtownsend reviewed (commented):
This looks really, really nice! I really only have a collection of smaller requests. The integration tests and updated documentation/examples are also great. I know we discussed on the community call the fact that the tests don't run/work, I will take a look to see if I can run them locally

brooksmtownsend commented:
I do reproduce the test failures on debug mode 😢

Mossaka commented on crates/provider-blobstore-nats/src/blobstore.rs:
```diff
@@ -0,0 +1,678 @@
+        &self,
+        context: Option<Context>,
+        id: bindings::wrpc::blobstore::types::ObjectId,
+        start: u64,
+        end: u64,
```
it doesn't seem like the `start` and `end` parameters are being used. See this one as an example: https://github.com/wasmCloud/wasmCloud/blob/main/crates/provider-blobstore-s3/src/lib.rs#L790

ossfellow replied:
NATS doesn't support reading ranges; that's why in all NATS providers the range is not being used.

Mossaka commented on crates/provider-blobstore-nats/src/blobstore.rs:
```diff
@@ -0,0 +1,678 @@
+            let (tx, rx) = mpsc::channel(16);
+            anyhow::Ok((
+                Box::pin(ReceiverStream::new(rx)) as Pin<Box<dyn Stream<Item = _> + Send>>,
+                Box::pin(async move {
```
Can this be simplified to one `async move`?

ossfellow replied:
When I was working on the provider, I had tried, but couldn't; [azure-blobstore-provider](https://github.com/wasmCloud/wasmCloud/blob/acdde189149b29732b31055924a97ceb68259d13/crates/provider-blobstore-azure/src/lib.rs#L449) has the same implementation.

Mossaka commented on crates/provider-blobstore-nats/src/blobstore.rs:
```diff
@@ -0,0 +1,678 @@
+            dst_container
+                .put(metadata, &mut src_object)
+                .await
+                .context("failed to move object to destination container")?;
```
Just want to confirm - if destination object exists, will this overwrite the destination object? 

Quote from the spec:

> overwrites destination object if it already existed.

ossfellow replied:
Yes, it does; that's NATS default behaviour.

Mossaka commented on crates/provider-blobstore-nats/src/blobstore.rs:
```diff
@@ -0,0 +1,678 @@
+                .map_err(|e| e.to_string())?;
+            // Delete the object
+            let result: Result<(), String> =
+                container.delete(id.object).await.map_err(|e| e.to_string());
```
Just want to make sure that if the object does not exist, this will **not** return an error

ossfellow replied:
it does; that's how [azure-blobstore-provider](https://github.com/wasmCloud/wasmCloud/blob/acdde189149b29732b31055924a97ceb68259d13/crates/provider-blobstore-azure/src/lib.rs#L360) has implemented it too.

thomastaylor312 reviewed (approved):
Hey there @ossfellow I had a few more small comments (mostly nits) and then there were some comments from @brooksmtownsend that hadn't been handled yet. Once those are all addressed, I think we're good to go!

thomastaylor312 reviewed (changes requested):
Whoops accidentally hit approve rather than request changes

ossfellow commented on .github/workflows/wasmcloud.yml:
```diff
@@ -316,6 +319,9 @@ jobs:
 
+          - name: blobstore-nats
+            subject: BLOBSTORE_NATS_SUBJECT
+
```
@brooksmtownsend, please note of this self-reminder, if it's still applicable, as I'm resolving this conversation.

ossfellow commented on crates/provider-blobstore-nats/README.md:
```diff
@@ -0,0 +1,56 @@
+
+| **Property**    | **Environment Variable** | **Description**                                                                    | **Default** |
+|:----------------|:------------------------|:----------------------------------------------------------------------------------|:------------|
+| `max_age`       | `CONFIG_NATS_STORAGE_MAX_AGE` | Maximum age of any blob in the container, expressed in seconds              | `315,569,520` (10 years) |
```
Thanks for the catch; I've made sure the code, comments and documentation are in sync, and state 0 (unlimited).

ossfellow commented on crates/provider-blobstore-nats/src/blobstore.rs:
```diff
@@ -0,0 +1,678 @@
+                Box::pin(async move {
+                    async move {
+                        // Stream the object data in chunks of 1024 bytes
+                        let mut buffer = vec![0; 1024];
```
Thanks for the hints; I improved how `chunk size` is set for object creation (256KB default, twice the NATS default), and how it's determined for reads (i.e. NATS object `size` divided by its number of `chunks`).

For object creation time, I took advantage of new fields in NATS object metadata to store the `created_at`, and later retrieve it (via NATS object info) when object metadata is requested.

ossfellow commented on crates/provider-blobstore-nats/src/config.rs:
```diff
@@ -0,0 +1,446 @@
+pub struct StorageConfig {
+    /// Maximum age of any object in the container, expressed in nanoseconds
+    #[serde(default)]
+    pub max_age: core::time::Duration,
```
As mentioned earlier, the code and documentation inconsistency has been addressed.

ossfellow commented:
@thomastaylor312, @brooksmtownsend, in the latest round of updates, in addition to what I mentioned in PR feedback responses, I took the hint @Mossaka had provided and treat deleting non-existent objects gracefully. I also added a new integration test for it.

ossfellow commented on crates/provider-blobstore-nats/src/blobstore.rs:
```diff
@@ -0,0 +1,678 @@
+                Box::pin(tokio_stream::wrappers::ReceiverStream::new(rx))
+                    as Pin<Box<dyn Stream<Item = Vec<String>> + Send>>,
+                Box::pin(async move {
+                    while let Some(object) = objects.next().await {
```
This has been improved to chunk object names as suggested.

lachieh reviewed (dismissed):
Smallest change, I know that this will trigger spelling errors in the docs build unless it is marked as code.

ossfellow commented on crates/provider-blobstore-nats/README.md:
```diff
@@ -0,0 +1,56 @@
+
+This capability provider is an implementation of the following interfaces of `wasi:blobstore` proposal, backed by NATS [Object Store](https://docs.nats.io/nats-concepts/jetstream/obj_store):
+
+- wasi:blobstore/blobstore
```
I'll do a documentation update PR tonight after work.

ossfellow commented:
> Smallest change, I know that this will trigger spelling errors in the docs build unless it is marked as code.

PR updated.

ossfellow commented:
@thomastaylor312, @brooksmtownsend, thank you for the reviews and help. I've rebased and squashed to commits into one.

ossfellow commented:
@thomastaylor312, thank you for the tip; it's done; the `async_nats` API has changed, and added 1headers` and `metadata`, which I've added, albeit without supplying any values.

lachieh commented:
🎉 Thanks to everyone involved with getting this across the line!
